File: 002. Add Two Numbers/add_two_numbers.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def addTwoNumbers(l1, l2):
    dummy = ListNode(0)
    current = dummy
    carry = 0

    while l1 or l2 or carry:
        val1 = l1.val if l1 else 0
        val2 = l2.val if l2 else 0

        total = val1 + val2 + carry
        carry = total // 10
        new_digit = total % 10

        current.next = ListNode(new_digit)
        current = current.next

        l1 = l1.next if l1 else None
        l2 = l2.next if l2 else None

    return dummy.next

def build_linked_list(values):
    head = ListNode(values[0])
    current = head
    for val in values[1:]:
        current.next = ListNode(val)
        current = current.next
    return head

def print_linked_list(node):
    vals = []
    while node:
        vals.append(str(node.val))
        node = node.next
    print(" -> ".join(vals))

if __name__ == "__main__":
    # Example 1: [2,4,3] + [5,6,4] = [7,0,8]
    l1 = build_linked_list([2,4,3])
    l2 = build_linked_list([5,6,4])
    result = addTwoNumbers(l1, l2)

    print("Result 1:")
    print_linked_list(result)

    # Example 2: [0] + [0] = [0]
    l1 = build_linked_list([0])
    l2 = build_linked_list([0])
    result = addTwoNumbers(l1, l2)

    print("Result 2:")
    print_linked_list(result)

    # Example 3: [9,9,9,9,9,9,9] + [9,9,9,9] = [8,9,9,9,0,0,0,1]
    l1 = build_linked_list([9,9,9,9,9,9,9])
    l2 = build_linked_list([9,9,9,9])
    result = addTwoNumbers(l1, l2)

    print("Result 3:")
    print_linked_list(result)

File: 002. Add Two Numbers/add_two_numbers.js
class ListNode {
    constructor(val, next = null) {
        this.val = val;
        this.next = next;
    }
}

function addTwoNumbers(l1, l2) {
    const dummy = new ListNode(0);
    let current = dummy;
    let carry = 0;

    while (l1 !== null || l2 !== null || carry !== 0) {
        const val1 = l1 ? l1.val : 0;
        const val2 = l2 ? l2.val : 0;
        const sum = val1 + val2 + carry;

        carry = Math.floor(sum / 10);
        const newDigit = sum % 10;

        current.next = new ListNode(newDigit);
        current = current.next;

        l1 = l1 ? l1.next : null;
        l2 = l2 ? l2.next : null;
    }

    return dummy.next;
}

function buildLinkedList(arr) {
    const head = new ListNode(arr[0]);
    let current = head;
    for (let i = 1; i < arr.length; i++) {
        current.next = new ListNode(arr[i]);
        current = current.next;
    }
    return head;
}

function printLinkedList(node) {
    const values = [];
    while (node !== null) {
        values.push(node.val);
        node = node.next;
    }
    console.log(values.join(" -> "));
}

function main() {

    // Example 1: [2, 4, 3] + [5, 6, 4] =[7, 0, 8]
    const l1A = buildLinkedList([2, 4, 3])
    const l2A = buildLinkedList([5, 6, 4])
    const resultA = addTwoNumbers(l1A, l2A)

    console.log("Result 1:")
    printLinkedList(resultA)

    // Example 2: [0] + [0] =[0]
    const l1B = buildLinkedList([0])
    const l2B = buildLinkedList([0])
    const resultB = addTwoNumbers(l1B, l2B)

    console.log("Result 2:")
    printLinkedList(resultB)

    // Example 3: [9, 9, 9, 9, 9, 9, 9] + [9, 9, 9, 9] =[8, 9, 9, 9, 0, 0, 0, 1]
    const l1C = buildLinkedList([9, 9, 9, 9, 9, 9, 9])
    const l2C = buildLinkedList([9, 9, 9, 9])
    const resultC = addTwoNumbers(l1C, l2C)

    console.log("Result 3:")
    printLinkedList(resultC)

}

main();
